### PR TITLE
Add ExpressionOptimizer to RowExpressionService

### PR DESCRIPTION
**Pull request by highker**
ExpressionOptimizer helps to simplify a RowExpression into a minimum
form. For example, `length(query) > 100` will be translated into
`GREATER_THAN(length(query), CAST(Slice{base=[B@47199eaa, address=47, length=3}))`;
After optimization, it will be `GREATER_THAN(length(query), 100)`

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java [hunk 1]
```diff
@@ -137,6 +137,29 @@ public RowExpressionInterpreter(RowExpression expression, Metadata metadata, Con
         this.visitor = new Visitor();
     }
 
+    public static final class Optimizer
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java]: Probably move it out as a standalone class?

**Comment by highker** [comment]
Comments addressed; thanks for review.

File: presto-spi/src/main/java/com/facebook/presto/spi/relation/ExpressionOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.relation;
+
+import com.facebook.presto.spi.ConnectorSession;
+
+public interface ExpressionOptimizer
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/relation/ExpressionOptimizer.java]: Curious: Why call it `ExpressionOptimizer` ? -- Now we have two `ExpressionOptimizer` right? (another in `com.facebook.presto.sql.relational.optimizer`)

File: presto-spi/src/main/java/com/facebook/presto/spi/relation/ExpressionOptimizer.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+{
+    /**
+     * Optimize a RowExpression to
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/relation/ExpressionOptimizer.java]: nit: unfinished masterwork :D

File: presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionOptimizer.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    public RowExpressionOptimizer(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public RowExpression optimize(RowExpression rowExpression, Level level, ConnectorSession session)
+    {
+        if (level == Level.SERIALIZABLE) {
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionOptimizer.java]: nit: I would use switch/case for enums.

**Pull request merged**